connector/github/github.go

```diff
@@ -380,43 +385,144 @@ func filterTeams(userTeams, configTeams []string) (teams []string) {
+// URLs, and returns true only if a "next" URL is found. The next pages' URL is
+// returned if a "next" URL is found. apiURL is returned if apiURL equals the
+// "last" URL or no "next" or "last" URL are found.
+//
+// https://developer.github.com/v3/#pagination
+func getPagination(apiURL string, resp *http.Response) (string, bool) {
+	if resp == nil {
+		return apiURL, false
+	}
+
+	links := resp.Header.Get("Link")
+	if len(reLast.FindStringSubmatch(links)) > 1 {
+		lastPageURL := reLast.FindStringSubmatch(links)[1]
+		if apiURL == lastPageURL {
+			return apiURL, false
 		}
-		return u, fmt.Errorf("%s: %s", resp.Status, body)
+	} else {
+		return apiURL, false
 	}
 
-	if err := json.NewDecoder(resp.Body).Decode(&u); err != nil {
-		return u, fmt.Errorf("failed to decode response: %v", err)
+	if len(reNext.FindStringSubmatch(links)) > 1 {
+		return reNext.FindStringSubmatch(links)[1], true
+	}
+
+	return apiURL, false
+}
+
+// user holds GitHub user information (relevant to dex) as defined by
+// https://developer.github.com/v3/users/#response-with-public-profile-information
+type user struct {
+	Name  string `json:"name"`
+	Login string `json:"login"`
+	ID    int    `json:"id"`
+	Email string `json:"email"`
+}
+
+// user queries the GitHub API for profile information using the provided client.
+//
+// The HTTP client is expected to be constructed by the golang.org/x/oauth2 package,
+// which inserts a bearer token as part of the request.
+func (c *githubConnector) user(ctx context.Context, client *http.Client) (user, error) {
+
+	// https://developer.github.com/v3/users/#get-the-authenticated-user
+	apiURL, resourcePath := c.apiURL, "/user"
+
+	var (
+		u   user
+		err error
+	)
+	if _, err = get(ctx, client, apiURL, resourcePath, &u); err != nil {
+		return u, err
+	}
+
+	// Only pulic user emails are returned by 'GET /user'. u.Email will be empty
+	// if a users' email is private. We must retrieve private emails explicitly.
+	if u.Email == "" {
+		if u.Email, err = c.userEmail(ctx, client); err != nil {
+			return u, err
+		}
 	}
 	return u, nil
 }
 
+// userEmail holds GitHub user email information as defined by
+// https://developer.github.com/v3/users/emails/#response
+type userEmail struct {
+	Email      string `json:"email"`
+	Verified   bool   `json:"verified"`
+	Primary    bool   `json:"primary"`
+	Visibility string `json:"visibility"`
+}
+
+// userEmail queries the GitHub API for a users' email information using the
+// provided client. Only returns the users' verified, primary email (private or
+// public).
+//
+// The HTTP client is expected to be constructed by the golang.org/x/oauth2 package,
+// which inserts a bearer token as part of the request.
+func (c *githubConnector) userEmail(ctx context.Context, client *http.Client) (string, error) {
+
+	// https://developer.github.com/v3/users/emails/#list-email-addresses-for-a-user
+	apiURL, resourcePath := c.apiURL, "/user/emails"
+
+	var (
```

Comment:
You're overusing this pattern. Probably better to just define things when they're actually used. e.g. within the for loop.